pkg/uio/buffer.go

```diff
@@ -0,0 +1,209 @@
+// Copyright 2018 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package uio
+
+import (
+	"encoding/binary"
+	"io"
+)
+
+// Marshaler is the interface implemented by an object that can marshal itself
+// into binary form.
+//
+// Marshal appends data to the buffer b.
+type Marshaler interface {
+	Marshal(b *Buffer)
+}
+
+// Unmarshaler is the interface implemented by an object that can unmarshal a
+// binary representation of itself.
+//
+// Unmarshal consumes data from the buffer b.
+type Unmarshaler interface {
+	Unmarshal(b *Buffer) error
+}
+
+// Buffer implements functions to manipulate byte slice buffers in specific
+// byte orders.
+type Buffer struct {
+	// data is the underlying data.
+	data []byte
+
+	// order is the byte order to write in / read in.
+	order binary.ByteOrder
+}
+
+// NewBEBuffer consumes b for marshaling or unmarshaling in big endian.
+func NewBEBuffer(b []byte) *Buffer {
+	return &Buffer{
+		data:  b,
+		order: binary.BigEndian,
+	}
+}
+
+// NewBuffer consumes b for marshaling or unmarshaling in the given byte order.
+func NewBuffer(b []byte, order binary.ByteOrder) *Buffer {
+	return &Buffer{
+		data:  b,
+		order: order,
+	}
+}
+
+// append appends n bytes to the Buffer and returns a slice pointing to the
+// newly appended bytes.
+func (b *Buffer) append(n int) []byte {
+	b.data = append(b.data, make([]byte, n)...)
+	return b.data[len(b.data)-n:]
+}
+
+// Data is unconsumed data remaining in the Buffer.
+func (b *Buffer) Data() []byte {
+	return b.data
+}
+
+// Remaining consumes and returns a copy of all remaining bytes in the Buffer.
+func (b *Buffer) Remaining() []byte {
+	p := b.Consume(len(b.Data()))
+	cp := make([]byte, len(p))
+	copy(cp, p)
+	return cp
+}
+
+// consume consumes n bytes from the Buffer. It returns nil, false if there
+// aren't enough bytes left.
+func (b *Buffer) consume(n int) ([]byte, bool) {
+	if !b.Has(n) {
+		return nil, false
+	}
+	rval := b.data[:n]
+	b.data = b.data[n:]
+	return rval, true
+}
+
+// Consume consumes n bytes from the Buffer. It returns nil if there aren't
+// enough bytes left.
+func (b *Buffer) Consume(n int) []byte {
+	v, ok := b.consume(n)
+	if !ok {
+		return nil
+	}
+	return v
+}
+
+// Has returns true if n bytes are available.
+func (b *Buffer) Has(n int) bool {
+	return len(b.data) >= n
+}
+
+// Len returns the length of the remaining bytes.
+func (b *Buffer) Len() int {
+	return len(b.data)
+}
+
+// Read implements io.Reader.Read.
+func (b *Buffer) Read(p []byte) (int, error) {
+	v, ok := b.consume(len(p))
+	if !ok {
+		return 0, io.ErrUnexpectedEOF
+	}
+	return copy(p, v), nil
+}
+
+// Read8 reads a byte from the Buffer.
+func (b *Buffer) Read8() uint8 {
```

Comment:
well ... can't really disinguish error from data here, is that desired?

Reply:
yeah, forgot to add the Error() function.